submissions/DmitryHniezdilov/task_js_oop/app.js

```diff
@@ -0,0 +1,296 @@
+};
+
+const Character = function (config, imgUrl) {
+  this.config = config;
+  this.sprite = imgUrl;
+};
+
+Character.prototype.endOfGame = function (result, config) {
+  const {
+    POPUP_MESSAGE_ACTIVE_CLASS,
+    ELEMENT_POPUP_MESSAGE,
+    ELEMENT_INFO_LEVEL,
+    ELEMENT_INFO_SPEED,
+    END_OF_GAME_WIN,
+    END_OF_GAME_LOSE,
+    PRINT_MESSAGE_WIN,
+    PRINT_MESSAGE_LOSE,
+    ENEMIES_CONF,
+    STATE,
+  } = config;
+
+  const updateMarkupInfo = (popupText) => {
+    ELEMENT_POPUP_MESSAGE.innerText = popupText;
+    ELEMENT_INFO_LEVEL.innerText = STATE.currentLevel;
+    ELEMENT_INFO_SPEED.innerText = `x${STATE.speedMultiplicator}`;
+    ELEMENT_POPUP_MESSAGE.classList.add(POPUP_MESSAGE_ACTIVE_CLASS);
+  };
+
+  isFreeze = true;
+  if (result === END_OF_GAME_WIN) {
+    STATE.currentLevel++;
+    STATE.speedMultiplicator = +(
+      STATE.speedMultiplicator + ENEMIES_CONF.speedIncrease
+    ).toFixed(1);
+    updateMarkupInfo(PRINT_MESSAGE_WIN);
+  } else if (result === END_OF_GAME_LOSE) {
+    STATE.currentLevel = 1;
+    STATE.speedMultiplicator = 1;
+    updateMarkupInfo(PRINT_MESSAGE_LOSE);
+  }
+
+  setTimeout(() => {
+    ELEMENT_POPUP_MESSAGE.classList.remove(POPUP_MESSAGE_ACTIVE_CLASS);
+    this.resetGame();
+    isFreeze = false;
+  }, 900);
+};
+
+const Enemy = function (x, y, minSpeed, maxSpeed, config) {
+  Character.call(this, config, config.ENEMIES_CONF.img);
+  this.ENEMIES_CONF = config.ENEMIES_CONF;
+  this.FIELD_WIDTH = config.FIELD_WIDTH;
+  this.BLOCK_WIDTH = config.BLOCK_WIDTH;
+  this.BLOCK_HEIGHT = config.BLOCK_HEIGHT;
+  this.x = x;
+  this.y = y;
+  this.speed = this.getRandomArbitrary(minSpeed, maxSpeed);
+};
+
+Enemy.prototype = Object.create(Character.prototype);
+
+Enemy.prototype.update = function (dt) {
+  if (!isFreeze) {
+    this.x += this.speed * dt;
+
+    if (this.x > this.FIELD_WIDTH) {
+      this.x = -this.BLOCK_WIDTH;
+    }
+
+    this.checkCollision();
+  }
+};
+
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.checkCollision = function () {
+  if (
+    player.y === this.y &&
+    player.x <= Math.floor(this.x) + this.BLOCK_WIDTH / 1.3 &&
+    player.x >= Math.floor(this.x) - this.BLOCK_WIDTH / 1.3
+  ) {
+    this.endOfGame(config.END_OF_GAME_LOSE, config);
+  }
+};
+
+Enemy.prototype.getRandomArbitrary = function (min, max) {
+  return Math.random() * (max - min) + min;
+};
+
+Enemy.prototype.startPositionOfEnemies = function (i) {
+  if (i === 1) {
+    return this.getRandomArbitrary(-this.BLOCK_WIDTH, this.BLOCK_WIDTH * 2);
+  } else if (i === 2) {
+    return this.getRandomArbitrary(-this.FIELD_WIDTH, -this.BLOCK_WIDTH);
+  } else {
+    return this.getRandomArbitrary(
+      -this.FIELD_WIDTH,
+      this.config.ORIGIN_COORDINATE_CHARACTERS.x
+    );
+  }
+};
+
+Enemy.prototype.createEnemies = function (config) {
+  const { ENEMIES_CONF, STATE, ORIGIN_COORDINATE_CHARACTERS, BLOCK_HEIGHT } =
```

Comment:
Looks like a factory pattern, but mixed with the class itself.
This could be a static method of the class. Otherwise it looks weird when we create an instance to create other instances.
If you are not sure of how to implement a static method in `prototype` syntax, just implement a global function outside class definition to create enemies.

Reply:
In order not to change the game control function Character.prototype.endOfGame, I initialized a config object in Character and created a function EnemiesFactory, which inherits this and .prototype from Enemy. After initializing the EnemiesFactory, I was able to use the enemy control methods as static methods (enemiesFactory.createEnemies();).
I haven't found a better way to update this code.